[inductor] Fix triton compiler error in multilayer any

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109325

Fixes #109196

When we have a split reduction and the tensor is not an even multiple of the split size,
we use `ops.masked` to pad to an even multiple. In the case here we generated:
```python
tmp5 = tl.where(mask, tmp4, 0)
```

which implicitly promotes our boolean value to `int32`. The fix is to give the default
value the same dtype as `result`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

peterbell10 committed:
[inductor] Fix triton compiler error in multilayer any

Fixes #109196

When we have a split reduction and the tensor is not an even multiple of the split size,
we use `ops.masked` to pad to an even multiple. In the case here we generated:
```python
tmp5 = tl.where(mask, tmp4, 0)
```

which implicitly promotes our boolean value to `int32`. The fix is to give the default
value the same dtype as `result`.

[ghstack-poisoned]

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -281,7 +281,13 @@ def index_expr(cls, expr, dtype):
-        return ops.where(new_mask, result, triton_constant(other))
+
+        # Take dtype from result to prevent accidental promotion
+        other = V.kernel.cse.generate(
+            V.kernel.compute,
+            f"tl.full({result}.shape, {triton_constant(other)}, {result}.dtype)",
+        )
```
Simply using `triton_constant(other)` should work, as when both inputs are in the same category, no broadcasting happens.

peterbell10 replied:
I'm not sure what you mean? The previous code already used `triton_constant`.

lezcano replied:
Ah, sorry, what I meant is that it would be enough to give it the right type. It could be simply by casting the `0` to `int`, `float`, `bool` depending on the type of `result`. But we don't have a way to know the dtype of result during compilation time so I guess this is alright.

We could also always use `False`, but this would be really odd, so nvm.

lezcano reviewed (approved):
Small nit, LGTM otherwise.

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge -r

peterbell10 committed:
Update on "[inductor] Fix triton compiler error in multilayer any"


Fixes #109196

When we have a split reduction and the tensor is not an even multiple of the split size,
we use `ops.masked` to pad to an even multiple. In the case here we generated:
```python
tmp5 = tl.where(mask, tmp4, 0)
```

which implicitly promotes our boolean value to `int32`. The fix is to give the default
value the same dtype as `result`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot rebase

peterbell10 committed:
Update on "[inductor] Fix triton compiler error in multilayer any"


Fixes #109196

When we have a split reduction and the tensor is not an even multiple of the split size,
we use `ops.masked` to pad to an even multiple. In the case here we generated:
```python
tmp5 = tl.where(mask, tmp4, 0)
```

which implicitly promotes our boolean value to `int32`. The fix is to give the default
value the same dtype as `result`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge -i

huydhn commented:
@pytorchbot drci

(Please ignore this, I's testing Dr.CI)
